### PR TITLE
server concept

### PR DESCRIPTION
**Pull request by chazkiker2**

### CODE REVIEW

**Comment by chazkiker2** [comment]
making this a PR so I can leave a review

File: src/client.ts [hunk 1]
```diff
@@ -1,46 +1,65 @@
...
-})
+type Options = { [key: string]: any };
+
+export class RyderWebsocketClient {
+	protected ws!: WebSocket;
+	protected options: Options;
+	protected connecting = false;
+
+	protected call_id = 0;
+	protected queue: Array<[Function, Function]> = [];
```
Comment (chazkiker2) [inline src/client.ts:12]: Two notes here: 

1. wondering if it may be worth re-working this into a separate class as we did with `train` in `RyderSerial.#train`. This can be a separate PR as well.
2. Let's avoid using the generic `Function` type and decide on a formal `Resolve` and `Reject` type.

File: src/client.ts [hunk 2]
```diff
@@ -1,46 +1,65 @@
...
+			}
+		});
+	}
+
+	close() {
+		this.connecting = false;
+		this.ws && this.ws.close();
+	}
+
+	async call(method: string, params: any[] = []) {
```
Comment (chazkiker2) [inline src/client.ts:46]: I think we should decide early on consistency for what type this is to return.

Some of the points made in [our `RyderSerial Response` discussion](https://github.com/Light-Labs/ryderserial-proto/pull/14#issuecomment-864350332) are probably relevant here as well

File: src/server/asyncrpcprovider.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+export class AsyncRpcProvider {
+	protected host;
+
+	constructor(host: any) {
+		this.host = host;
+	}
+	async call(request: RpcRequest): Promise<RpcResponse> {
+		const { id, method, params } = request || {};
+		if (!this.host.hasOwnProperty(method))
+			return { jsonrpc: '2.0', id: id!, error: { code: 0, message: 'unknown method' } };
```
Comment (chazkiker2) [inline src/server/asyncrpcprovider.ts:32]: when you include `id: id!` are you hoping for this to fail out if `id` is `undefined` or not? b/c it's currently just telling TypeScript that the id will NOT be undefined. so no check will happen here at runtime, essentially.

as it stands currently, it looks like `const {id, method, params} = request || {};` is allowing for that to be undefined if no request exists? (which would throw in TypeScript as it is...)

File: src/server/index.ts [hunk 4]
```diff
@@ -1,93 +1,43 @@
...
+	}
+
+	listen(config: Config) {
+		this.wss = new WebSocket.Server({ host: config.host, port: config.port });
+		this.emit('listening');
+		this.wss.on('connection', this.handle_connection.bind(this));
+	}
+
+	protected handle_connection(client: WebSocket, request: IncomingMessage) {
+		if (!['127.0.0.1', '::1'].includes(request.socket.remoteAddress!))
```
Comment (chazkiker2) [inline src/server/index.ts:26]: once again wondering if you're hoping to fail out here if `request.socket.remoteAddress` is `undefined` or not? This line is telling TypeScript that `remoteAddress` will definitively NOT be `undefined` at run time.

Thus we won't `client.close(1)` if `request.socket.remoteAddress` is undefined.

File: src/stacks_wallet_provider.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
+/*
```
Comment (chazkiker2) [inline src/stacks_wallet_provider.ts:1]: big fan of what these examples are looking like

Comment (chazkiker2) [reply src/stacks_wallet_provider.ts:1]: this will make for a really easy client SDK as well if we ever wanted to do that.

**Review by chazkiker2 — COMMENTED** [review]
@MarvinJanssen I definitely like the direction this is headed. 

Left a couple nits and comments. 

But generally I think we should build off of this approach. I think we will benefit from formalizing types earlier rather than later. And these are the moments I do wish `RyderSerial` responded with a consistent response type... we're going to end up having a bunch of `Promise<number | string>` all over the place. 

Also (easy) we'll need to `npm i --save-dev @types/ws`

File: src/server/index.ts [hunk 6]
```diff
@@ -1,93 +1,43 @@
...
+			return client.close(1);
+		this.emit('connection', client);
+		client.on('message', this.handle_message.bind(this));
+	}
+
+	protected handle_message(client: WebSocket, data: WebSocket.Data) {
```
Comment (chazkiker2) [inline src/server/index.ts:32]: `type WebSocket.Data = string | Buffer | ArrayBuffer | Buffer[]`

are you returning out with the `if (typeof data !== "string") return;` b/c we don't want to handle `Buffer | ArrayBuffer | Buffer[]`? And if so, should we just make `data` parameter `string`?

File: src/client.ts (hunk 1 above)
Comment (MarvinJanssen) [reply src/client.ts:12]: Not sure if it's necessary because it'll always be the same two functions pushed onto the queue. I'll leave it up to you, as I don't know if there's a more formal TS way to represent resolve and reject.

File: src/client.ts (hunk 2 above)
Comment (MarvinJanssen) [reply src/client.ts:46]: Great point. We can also debate making `call()` protected to force devs to use the specific call implementations.

File: src/server/asyncrpcprovider.ts (hunk 3 above)
Comment (MarvinJanssen) [reply src/server/asyncrpcprovider.ts:32]: `id` is optional, only specified is a response is expected. I just didn't want TS to complain. We can probably just call the method and discard the result if `id` is undefined.

File: src/server/index.ts (hunk 4 above)
Comment (MarvinJanssen) [reply src/server/index.ts:26]: Doesn't the `!` simply silence that particular TS check? It'll just compile down into the same JS and thus the if statement fails because `undefined` is not in the array. Is my assumption wrong?

(I didn't actually check what JS it compiles down into.)

File: src/stacks_wallet_provider.ts (hunk 5 above)
Comment (MarvinJanssen) [reply src/stacks_wallet_provider.ts:1]: Yeah, we will probably have to lead the way and get something like this adopted in `stacks-wallet-web`.

File: src/server/index.ts (hunk 6 above)
Comment (MarvinJanssen) [reply src/server/index.ts:32]: It should always be a string if we're sending JSON. Since it's JSON-RPC over ws I figured to reject the others immediately because it means the sender is doing something funny. Still, we can just read the buffers as UTF-8 strings and and parse them as JSON. What do you think?

**Comment by MarvinJanssen** [comment]
Thanks for the review. I know it's still in the concept phase but these are some great thoughts.

File: src/server/index.ts (hunk 6 above)
Comment (chazkiker2) [reply src/server/index.ts:32]: gotchya that makes sense — let's leave it for now but maybe later incorporate some sort of error message.

File: src/server/index.ts (hunk 4 above)
Comment (chazkiker2) [reply src/server/index.ts:26]: Oh yes I reversed that in my head... this WILL close if `request.socket.remoteAddress` is `undefined`...  which makes sense

File: src/server/asyncrpcprovider.ts (hunk 3 above)
Comment (chazkiker2) [reply src/server/asyncrpcprovider.ts:32]: looks like `id` is required in `RpcResponseBase` but we can make it optional for sure. 

And then you wouldn't need to put in the `!` since TS will know it's optional (and we can just shorthand it like 
```typescript
{ jsonrpc: '2.0', id, result };
```

File: src/client.ts (hunk 1 above)
Comment (chazkiker2) [reply src/client.ts:12]: For `Resolve` and `Reject` types (if you're returning `Promise<void>`)

```typescript
type Resolve = (value: void | PromiseLike<void>) => void;
type Reject = (reason?: any) => void;
```

But I'm fine with leaving `queue` as a simple array, actually (not its own implementation)

File: src/server/asyncrpcprovider.ts (hunk 3 above)
Comment (MarvinJanssen) [reply src/server/asyncrpcprovider.ts:32]: It's required in the response but not the request. We should just not send anything back if the `id` is missing.

File: src/server/index.ts (hunk 6 above)
Comment (MarvinJanssen) [reply src/server/index.ts:32]: Yeah sending a JSON-RPC error object makes sense.
